### PR TITLE
Need to be able to create structs via macro_rules

### PR DESCRIPTION
**Pull request by gilescope**
(Currently only possible using tt escape type which doesn't do any processing.)

There's quite a few places I think we need to update but using `pyclass` as an example to agree the right way to do things. When you pass the argument via macro_rules unless it's the escape sequence type of `tt` then the Expr::Path seems to get wrapped in an Expr::Group whatever type I choose for the variable I'm passing in: `ident`, `path`, `expr`.

(btw I notice you can't match on box which might make the code a tad nicer.)

### CODE REVIEW

**Comment by gilescope** [comment]
(The test fails to compile without the patch which is why there's no assertions in it)

File: pyo3-derive-backend/src/pyclass.rs
Comment (kngwyu) [inline pyo3-derive-backend/src/pyclass.rs]: `expected` macro does not accept formatting args. Just `type name` would be sufficient.
And this can be simplified like
```suggestion
                // For macro args(e.g., $name)
                syn::Expr::Group(exp_grp) => match &*exp_grp.expr {
                    syn::Expr::Path(exp) if exp.path.segments.len() == 1 => {
                        self.name = Some(exp.clone().into());
                    }
                    _ => expected!("type name (e.g., Name)"),
                },
```

File: tests/test_macros.rs
Comment (kngwyu) [inline tests/test_macros.rs]: I don't want to have too many test files.
Could you please move this test to `test_class_new` or so?

**Comment by kngwyu** [comment]
Thanks :+1:

File: tests/test_macros.rs
Comment (gilescope) [reply tests/test_macros.rs]: Hold fire on that thought. I will add in the additional tests and we can agree where to put them once I’ve made sure we can macro_rules create anything with an attribute.

**Comment by programmerjake** [comment]
Encountered this issue myself.

One suggestion: maybe there should be an `unwrap_group` function that unwraps a group or returns the input unchanged:

```rust
// pseudo-code, probably needs some derefs and other misc. cleanup
fn unwrap_group(expr: &syn::Expr) -> &syn::Expr {
    if let syn::Expr::Group(g) = expr {
        &*g.expr
    }
    expr
}
```

that way, the `match` cases can be simplified:
```rust
            "name" => match unwrap_group(&**right) {
                syn::Expr::Path(exp) if exp.path.segments.len() == 1 => {
                    self.name = Some(exp.clone().into());
                }
                _ => expected!("type name (e.g., Name)"),
            },
```

**Comment by scalexm** [comment]
Btw, I would suggest a `while let Group(g) = expr` approach rather than unwrapping a single layer. Not sure if that could actually happen but it’s harmless.

**Comment by davidhewitt** [comment]
👍 also my parsing improvements in #1567 might resolve a lot of the problems here. I plan to retest this once that PR is merged.

**Comment by davidhewitt** [comment]
As this branch is very old, I just rebased it and will merge it if tests are green 🤞

**Comment by scalexm** [comment]
More of these problems started to appear with Rust 1.53 (at first I thought it was a compiler régression, but actually it’s just the compiler adding a few more groups and pyo3 not handling them correctly).

I think we should use `unwrap_group` quite liberally: I have found a few more places where we need it. Once this is merged, I’ll open a follow-up PR.

**Review by messense — APPROVED** [review]

**Pull request merged**
